### PR TITLE
add sclae_log parameters for quantization.

### PR DESCRIPTION
**Pull request by krodyush**
add sclae_log parameters for quantization. its allow to increse convergence speed for high scales and increase accuracy for low scales.

### CODE REVIEW

File: nncf/quantization/layers.py [hunk 1]
```diff
@@ -230,17 +240,40 @@ def __init__(self, config):
         self.collect_scale_statistics = False
         if self.per_channel:
             self.scale_shape = get_per_channel_scale_shape(self.input_shape, self.is_weights)
-        self.scale = nn.Parameter(torch.ones(self.scale_shape), requires_grad=True)
-        self.eps = 1e-16
+        self.scale_tensor = nn.Parameter(torch.ones(self.scale_shape), requires_grad=True)
```
Comment (vshampor) [inline nncf/quantization/layers.py:243]: Mark the internal parameter with a leading underscore. `_scale_tensor`

File: nncf/quantization/layers.py [hunk 2]
```diff
@@ -230,17 +240,40 @@ def __init__(self, config):
...
+        if self.scale_log_flag:
```
Comment (vshampor) [inline nncf/quantization/layers.py:244]: This flag should also be marked as internal - the logic does not support changing it externally.

File: nncf/quantization/layers.py [hunk 3]
```diff
@@ -230,17 +240,40 @@ def __init__(self, config):
...
+            self.scale_tensor.data.log_()
+            self.eps = 0
+        else:
+            self.eps = 1e-16
         if config.signedness_to_force is not None:
             self.signed = int(config.signedness_to_force)
         self.set_level_ranges()
 
+        #register hooks to convert scale from log vertion to original vertion during load and save
+        def hook_fn_load_state_dict(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
```
Comment (vshampor) [inline nncf/quantization/layers.py:254]: How about redefining the `torch.nn.Module._load_state_dict` and `torch.nn.Module._state_dict` functions instead of using the hook mechanism?

File: nncf/quantization/layers.py [hunk 4]
```diff
@@ -230,17 +240,40 @@ def __init__(self, config):
...
+        def hook_fn_state_dict(module, destination, prefix, local_metadata):
+            v = destination.pop(prefix + 'scale_tensor')
+            if module.scale_log_flag:
+                v = v.exp().detach()
+            destination[prefix + 'scale'] = v
+        self.state_dict_hook = HookAutoRemove(self._register_state_dict_hook(hook_fn_state_dict))
+
+    @property
+    def scale(self):
+        return self.scale_tensor.exp() if self.scale_log_flag else self.scale_tensor
```
Comment (vshampor) [inline nncf/quantization/layers.py:270]: Should define a setter as well.

**Review by vshampor — CHANGES_REQUESTED** [review]
The main idea is correct, but there are still some changes that should be made here.

File: nncf/quantization/layers.py (hunk 3 above)
Comment (krodyush) [reply nncf/quantization/layers.py:254]: Orignal BaseQuantizer (see top of the file) use the hook. I just used the same approach for consistency.

File: nncf/quantization/layers.py (hunk 4 above)
Comment (krodyush) [reply nncf/quantization/layers.py:270]: I can define setter but it is not clear what input should be for this setter. And where it will be used.

File: nncf/quantization/layers.py (hunk 1 above)
Comment (krodyush) [reply nncf/quantization/layers.py:243]: ok

File: nncf/quantization/layers.py (hunk 2 above)
Comment (krodyush) [reply nncf/quantization/layers.py:244]: ok

File: nncf/quantization/layers.py (hunk 4 above)
Comment (krodyush) [reply nncf/quantization/layers.py]: ok. I prepare and commit some variant for scale. if it is ok lets keep it. But still have no idea why it is needed.

File: nncf/quantization/layers.py (hunk 3 above)
Comment (ljaljushkin) [reply nncf/quantization/layers.py:254]: Don't think it's a good idea to redefine standard torch functions. The behavior can be changed in the newer version, and our redefined implementation would lead to errors and bugs. API with hooks is the way, suggested and supported by PyTorch, to extend `load_state_dict` functionality. IMHO, it is better to stick to this approach

File: nncf/quantization/layers.py [hunk 5]
```diff
@@ -285,9 +330,11 @@ def apply_minmax_init(self, min_values, max_values, log_module_name: str = None)
 
         abs_max = torch.max(torch.abs(max_values), torch.abs(min_values))
         SCALE_LOWER_THRESHOLD = 0.1
-        self.scale.fill_(SCALE_LOWER_THRESHOLD)
+        #self._scale_tensor.fill_(SCALE_LOWER_THRESHOLD)
```
Comment (vshampor) [inline nncf/quantization/layers.py:333]: A bug?

**Review by vshampor — CHANGES_REQUESTED** [review]
I'll start the precommit so that we can see if this breaks anything. In the meanwhile, @krodyush, please add tests to check that the checkpoints are interchangeable between the log-scale and non-log-scale models.

Also, just to confirm - will the existing checkpoints have to be converted to work with the new code or not?

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/quantization/layers.py (hunk 5 above)
Comment (krodyush) [reply nncf/quantization/layers.py:333]: this line does not make sense. it fills the tensor that is fill again by next 2 lines

**Comment by krodyush** [comment]
> I'll start the precommit so that we can see if this breaks anything. In the meanwhile, @krodyush, please add tests to check that the checkpoints are interchangeable between the log-scale and non-log-scale models.
I have no any experience in your test system. if you give me draft code I could chnage it to have such testing 

> 
> Also, just to confirm - will the existing checkpoints have to be converted to work with the new code or not?
it will

File: nncf/quantization/layers.py (hunk 5 above)
Comment (krodyush) [reply nncf/quantization/layers.py:333]: probably I wrong. you have to check with author of this code

**Comment by vshampor** [comment]
> > I'll start the precommit so that we can see if this breaks anything. In the meanwhile, @krodyush, please add tests to check that the checkpoints are interchangeable between the log-scale and non-log-scale models.
> > I have no any experience in your test system. if you give me draft code I could chnage it to have such testing
> 

Our test system is regular pytest. Pick any file from `tests/*.py` for reference.

> > Also, just to confirm - will the existing checkpoints have to be converted to work with the new code or not?
> > it will

Please add a script to convert the existing checkpoints to the new format under `tools`.

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by krodyush** [comment]
> > > Also, just to confirm - will the existing checkpoints have to be converted to work with the new code or not?
> > > it will
> 
> Please add a script to convert the existing checkpoints to the new format under `tools`.
Sorry. I did not carefully read you question first time. 
It is not required to convert old checkpoints.
There is no difference between old checkpoints and new checkpoints. model.state_dict() return same parameters.

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by krodyush** [comment]
Jenkins please retry a build

**Comment by krodyush** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by krodyush** [comment]
I see the fail that does not look as related to the scale_log

http://213.221.44.203/job/nncf/job/precommit/job/NNCF/650/consoleFull
11:51:20  =================================== FAILURES ===================================
11:51:20  _ test_pretrained_model_eval[classification-squeezenet1_1_cifar10_rb_sparsity_int8.json-cifar10-256-distributed] _
11:51:20  
11:51:20  config = {'batch_size': 256, 'dataset_path': '/mnt/pre_commit_data', 'model_name': 'squeezenet1_1_custom', 'nncf_config': {'com...ue}}], 'dataset': 'cifar10', 'input_info': {'sample_size': [1, 3, 32, 32]}, 'model': 'squeezenet1_1_custom', ...}, ...}
11:51:20  tmp_path = PosixPath('/tmp/pytest-of-root/pytest-0/test_pretrained_model_eval_cla0')
11:51:20  multiprocessing_distributed = True
11:51:20  
11:51:20      @pytest.mark.parametrize(" multiprocessing_distributed",
11:51:20                               (True, False),
11:51:20                               ids=['distributed', 'dataparallel'])
11:51:20      def test_pretrained_model_eval(config, tmp_path, multiprocessing_distributed):
11:51:20          config_factory = ConfigFactory(config['nncf_config'], tmp_path / 'config.json')
11:51:20          args = {
11:51:20              "--mode": "test",
11:51:20              "--data": config["dataset_path"],
11:51:20              "--config": config_factory.serialize(),
11:51:20              "--log-dir": tmp_path,
11:51:20              "--batch-size": config["batch_size"] * torch.cuda.device_count(),
11:51:20              "--workers": 0,  # Workaround for the PyTorch MultiProcessingDataLoader issue
11:51:20          }
11:51:20      
11:51:20          if multiprocessing_distributed:
11:51:20              args["--multiprocessing-distributed"] = None
11:51:20      
11:51:20          runner = Command(create_command_line(args, config["sample_type"]))
11:51:20          res = runner.run()
11:51:20  >       assert res == 0
11:51:20  E       assert 1 == 0
11:51:20  
11:51:20  test_sanity_sample.py:229: AssertionError
11:51:20  ----------------------------- Captured stdout call -----------------------------
11:51:20  /home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/jsonschema-3.2.0-py3.6.egg/jsonschema/validators.py:928: DeprecationWarning: The metaschema specified by $schema was not found. Using the latest draft to validate, but this will raise an error in the future.
11:51:20    cls = validator_for(schema)
11:51:20  Traceback (most recent call last):
11:51:20    File "<string>", line 1, in <module>
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
11:51:20      exitcode = _main(fd)
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 115, in _main
11:51:20      self = reduction.pickle.load(from_parent)
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/addict-2.4.0-py3.6.egg/addict/addict.py", line 33, in __setitem__
11:51:20      isFrozen = (hasattr(self, '__frozen') and
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/common/sample_config.py", line 77, in __getattr__
11:51:20      raise KeyError("Key {} not found in config".format(item))
11:51:20  KeyError: 'Key __frozen not found in config'
11:51:20  Traceback (most recent call last):
11:51:20    File "<string>", line 1, in <module>
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
11:51:20      exitcode = _main(fd)
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 115, in _main
11:51:20      self = reduction.pickle.load(from_parent)
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/addict-2.4.0-py3.6.egg/addict/addict.py", line 33, in __setitem__
11:51:20      isFrozen = (hasattr(self, '__frozen') and
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/common/sample_config.py", line 77, in __getattr__
11:51:20      raise KeyError("Key {} not found in config".format(item))
11:51:20  KeyError: 'Key __frozen not found in config'
11:51:20  Traceback (most recent call last):
11:51:20    File "<string>", line 1, in <module>
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
11:51:20      exitcode = _main(fd)
11:51:20    File "/usr/lib/python3.6/multiprocessing/spawn.py", line 115, in _main
11:51:20      self = reduction.pickle.load(from_parent)
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/addict-2.4.0-py3.6.egg/addict/addict.py", line 33, in __setitem__
11:51:20      isFrozen = (hasattr(self, '__frozen') and
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/common/sample_config.py", line 77, in __getattr__
11:51:20      raise KeyError("Key {} not found in config".format(item))
11:51:20  KeyError: 'Key __frozen not found in config'
11:51:20  Traceback (most recent call last):
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/classification/main.py", line 539, in <module>
11:51:20      main(sys.argv[1:])
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/classification/main.py", line 100, in main
11:51:20      start_worker(main_worker, config)
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/src/examples/common/execution.py", line 113, in start_worker
11:51:20      mp.spawn(main_worker, nprocs=config.ngpus_per_node, args=(config,))
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/torch-1.5.0-py3.6-linux-x86_64.egg/torch/multiprocessing/spawn.py", line 200, in spawn
11:51:20      return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/torch-1.5.0-py3.6-linux-x86_64.egg/torch/multiprocessing/spawn.py", line 158, in start_processes
11:51:20      while not context.join():
11:51:20    File "/home/jenkins/workspace/nncf/precommit/NNCF/venv/lib/python3.6/site-packages/torch-1.5.0-py3.6-linux-x86_64.egg/torch/multiprocessing/spawn.py", line 113, in join
11:51:20      (error_index, exitcode)
11:51:20  Exception: process 2 terminated with exit code 1
11:51:20  Process returncode = 1

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by krodyush** [comment]
@vshampor, quantization/test_precision_init.py::test_disable_quantizer_gradients was failed brequires it requires that module.scale has to be torch.nn.Parameter to be int module.named_parameters() list. At the same time you require that module.scale has to be property. I can not make both true.

**Comment by vshampor** [comment]
@krodyush you should probably just update the corresponding .json reference files to conform to the new convention.

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
@krodyush please rebase your changes upon the latest develop branch so that I could run the checkpoint evaluation jobs.

**Comment by vshampor** [comment]
Jenkins please retry a build

File: tests/quantization/test_unified_scales.py [hunk 6]
```diff
@@ -561,7 +561,8 @@ def test_unified_scales_are_identical_in_onnx(tmp_path):
 
     with torch.no_grad():
         for quant_info in compression_ctrl.non_weight_quantizers.values():
-            quant_info.quantizer_module_ref.scale *= torch.abs(torch.rand_like(quant_info.quantizer_module_ref.scale))
+            scale = quant_info.quantizer_module_ref.scale
+            quant_info.quantizer_module_ref.scale = scale*torch.abs(torch.rand_like(scale))
```
Comment (vshampor) [inline tests/quantization/test_unified_scales.py:565]: The idea behind the property usage was that all existing `.scale`-using code would still function properly with the change. Why didn't this line of code work?

File: nncf/quantization/layers.py [hunk 7]
```diff
@@ -285,9 +343,13 @@ def apply_minmax_init(self, min_values, max_values, log_module_name: str = None)
...
```
Comment (vshampor) [inline nncf/quantization/layers.py:346]: Please undo this change if you are not sure about it then.

**Review by vshampor — CHANGES_REQUESTED** [review]
The code seems to be working fine - the old checkpoints are loading and produce expected quality. Still, a few minor issues to fix before we merge this.

File: tests/quantization/test_unified_scales.py (hunk 6 above)
Comment (krodyush) [reply tests/quantization/test_unified_scales.py:565]: I already mentioned that torch.nn.Parameter does not work with property because torch.Module intercept this and raise error.  .scale=torch.nn.Parameter() will raise error that .scale already exist.

File: nncf/quantization/layers.py [hunk 8]
```diff
@@ -219,6 +226,13 @@ def enable_quantizers(self):
         self.originally_disabled = []
 
 
+class HookAutoRemove:
+    def __init__(self, hook):
+        self.hook = hook
+    def close(self):
```
Comment (ljaljushkin) [inline nncf/quantization/layers.py:232]: AFAIK, `close` should work context manager only. It seems you to override `__del__` method

Comment (krodyush) [reply nncf/quantization/layers.py]: You are right. I 've made blind copy from LoadStateListener class.
Thanks a lot!

**Comment by vshampor** [comment]
@krodyush since you disallowed editing this PR, I opened #329 with what I hope are improvements on top of your code. Please review that PR and see if it still does what you need it to do.

**Comment by krodyush** [comment]
"@krodyush since you disallowed editing this PR"
strange to see this. I believe that I have made changes for all you requests.
